### ISSUE TITLE
Feature Requests on Tokenizers - Serialization / Model-Fitting / Custom-Vocabulary

### ISSUE THREAD

**Issue by pandrey-fr** [tensorflow/text#417]
Hello,

I have been starting using TensorFlow Text in reaction to the deprecation of the TextEncoder classes in TensorFlow Datasets, with the hope that I could easily convert my code to switch from the latter to the former. Long story short: it proves hard.

There are some really exciting features in TensorFlow Text, and I would like to thank everybody who has been working on that, but as a user, after a few hours of testing things and reading the python part of the source code, I feel like a few things are lacking to make their adoption simple. So, I would like to humbly provide some feedback and request some features as to the `Tokenizer` classes, for the implementation of which I would actually be willing to give a hand.

#### (0) Tokenizer Serialization

It would be really nice to be able to dump and reload a Tokenizer without having to ponder about all the (very cool, modularity-yielding) instantiation arguments. I believe that the easiest would probably just be to mimic the `get_config` / `from_config` keras API, basically returning dict of serializable keyword arguments.

To be fair, my main issue at the moment is with the SentencepieceTokenizer (as well as the Wordpiece and Bert ones), whose main instantiation argument is made inaccessible after intantiation (at least using the publicly exposed attributes and methods). Being able to access the proto of a built tokenizer and export it back (either as bytes or using a hex-encoding trick to be writable to json) would be really nice.

#### (1) Tokenizer Fitting

At the moment, if I want to use a SentencepieceTokenizer on my data, I have to first use the `sentencepiece` module, use it to fit a model on my data (which is semi-easy: the tool is amazing, but the documentation is so split apart that it can take time to manage using it), read one of the output files's bytes (the 'proto', which is not clearly documented to be so) and instantiate a tokenizer from it (which to be built again will require remembering what the 'proto', but that is my former point).

I think that it would be neat to have a `SentencepieceTokenizer.build_from_data` method (on the model of `SubwordTextEncoder.build_from_data`) that would do the job of interfacing the `sentencepiece` package, gathering the proto and (last but not least) documenting the former package's options (at least the main ones). This could then serve for future implementations supporting alternative tokenizer-fitting algorithms, including simple no-subwords frequency-based ones (which are my last point).

#### (2) Custom-Vocabulary Tokenizers

Finally, it would be nice to be able to instantiate a tokenizer that works using an arbitrary input vocabulary (like the former `TokenTextEncoder` one). I know that this would be partly redundant with `tf.keras.layers.preprocessing.experimental.TextVectorization`, but it probably would not be hard to achieve, and would offer a non-layer alternative to the former. I am actually wondering to what extent a SentencepieceTokenizer subclass without sampling options could simply be used (with a wrapper to convert input vocabulary list to the proper proto format).

---

I am aware that building TensorFlow Text takes time, and I apologize if this is is any way redundant with existing roadmaps that I would not be aware of. I am truly looking forward to making TF-Text part of my tools, and am simply hoping that feedback and possible contribution through PRs (once the former points have been discussed, adjusted and validated) can help avoid having users (myself including) forced to write custom framework hacks that end up not being easily sharable nor maintainable.

Thank you for any time and effort put in reviewing what precedes. I hope it can result in actions, and am looking forward to contribute if needed.

**Comment by broken** [comment]
Thanks for your suggestions! I'll respond to each individually.

(0) For this, are you asking to have access to the instantiation parameters? In the case of SentencepieceTokenizer, this would be the bytes model you passed in? I'm not quite sure I understand why you would want this if you already have handles to that data when you instantiated the tokenizer.

(1) This is a current deficiency that we plan on addressing for the WordpieceTokenizer & SentencepieceTokenizer. We are actively building out a TFX / tf.Transform component that will fit vocabs for these, and plan on an "in-graph" solution like you describe as well.

(2) We have been discussing something similar to this. Originally, we wanted to provide just the core pieces so anybody could do what they wanted. Like for TextVectorization, you can do similar things by stringing many of our ops together - normalize_utf8, WhitespaceTokenizer, [StaticVocabularyTable](https://www.tensorflow.org/api_docs/python/tf/lookup/StaticVocabularyTable), ngrams, etc. However, we're coming to the realization there are many little decisions in there that most people really don't care about or need to care about (ie. NFKC or NFD normalization? Whitespace or UnicodeScript tokenization?) So, we are discussing what should be provided in this packaged tokenizer that just does what people want without having to think about the mundane details. I'll bring your comments up in our next meeting.

**Comment by pandrey-fr** [comment]
Hello,

Thank you very much for your answer! Here are mine in return:

(0-a) You have a fair point on the fact that someone who handles the proto can be expected to still be able to do so in the future; I was merely thinking of the (small) practicality gain of being able to store both the model proto and the other parameters in a single file. I was also thinking about the case when the model is built through a method that does not require the user to care about the source file; then, I think having a small (de)serialization API could be useful.

(0-b) On another matter (expanding my initial point), for SentencePiece, there are quite a lot of parameters serialized within the proto that are not publicly available in the current TF-Text wrapper (as opposed to that from the sentencepiece package); I am notably thinking about `eos_id`, `eos_piece` and the likes. This can make model handling slightly less efficient for users, because it means that one has to remember - or manually test - whether eos/bos are defined and how; it can also generate segmentation faults if one requires eos addition while that token is not defined. Hence my (new) point is that it could be useful to re-expose some parameters currently handled "in the back" (on the side, sanitization of inputs and/or parameters could be nice, to avoid segfaults interrupting python - it could be nice to prevent the user from asking the impossible, and/or to trim off oov input ids to functions like id_to_string and detokenize).

(1) I am very glad to learn about the planned in-graph solution and TFX components! I will therefore simply keep on relying on my verbose wrapper of sentencepiece.SentencePieceTrainer in the meanwhile, and am looking forward to deprecating it in favour of these upcoming tools.

(2) I appreciate the fact that this is being discussed; and also acknowledge how difficult it is to make design choices on this matter, for the reasons you mention. Thank you very much for you work and effort :)
